### PR TITLE
fix ICE when parsing lifetime as function argument

### PR DESCRIPTION
**Pull request by compiler-errors**
I don't really like this, but we basically need to emit an error instead of just delaying an bug, because there are too many places in the AST that aren't covered by my previous PRs...

cc: https://github.com/rust-lang/rust/issues/93282#issuecomment-1028052945

### CODE REVIEW

**Review by jackh726 — CHANGES_REQUESTED** [review]
This doesn't quite seem right :)

I would have expected that the "recovery" here is just to fall back to generics with a lifetime.

**Comment by compiler-errors** [comment]
> I would have expected that the "recovery" here is just to fall back to generics with a lifetime.

@jackh726, this fix is actually for when we parse lifetimes in the expression position. This happens when we have a lifetime token as the first token when parsing a loop expression.

So the deal is that we delayed a bug in the code, but none of the paths that I've previously added which do recovery for bad generics apply to something like:

```rust
fn main() {
  function('a,
        // ^^- here
```

File: compiler/rustc_typeck/src/check/expr.rs [hunk 1]
```diff
@@ -865,14 +865,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
                 }) => {
-                    // We have a situation like `while Some(0) = value.get(0) {`, where `while let`
-                    // was more likely intended.
-                    err.span_suggestion_verbose(
-                        expr.span.shrink_to_lo(),
-                        "you might have meant to use pattern destructuring",
-                        "let ".to_string(),
-                        Applicability::MachineApplicable,
-                    );
+                    // Check if our lhs is a child of the condition of a while loop
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/expr.rs:868]: Ok, this definitely got included from another PR of mine. I'll revert this!

**Comment by jackh726** [comment]
For the function arg case, I could see this as acceptable (given that you actually *can* use a labeled expr as an argument - or at least I would expect that to work). For the generics though, I'm a bit surprised we get to `parse_labeled_expr` in the first place. *Maybe* consts could have a labeled expr as a value (but I would expect those to have to be enclosed in `{}`). This can probably be conditional on if we're enclosed in `()` vs `{}`

**Comment by compiler-errors** [comment]
Well in the case we have an expr `x < 'a`, we need to parse that `'a` as an expression (just the loop label). This is important for the recovery in the case we do have a bad turbofish. But in the case we don't have a bad turbofish, we need to emit an error. 

We can't just raise a `PResult::Err` because we have no way of distinguishing that the error is one we can recover as turbofish generics.

It's just because the recovery is split over too many layers, since that `<` looks like an expression until _much_ later, when we see `>`, and we try to recover the chained comparison operator error.

**Comment by compiler-errors** [comment]
I think that explanation of mine is a bit complicated. To put it in a different way:

We don't start recovering a bad turbofish until _after_ we have parsed the bad label expression. If we continue to delay a bug in the label parsing code like we're doing now, we must be absolutely certain we're going to emit another bug later on all codepaths we took to get here, in order to prevent an ICE (sorry, you definitely know this already, just want to spell out my thought process clearly).

It turns out this doesn't always work out, since we cannot bubble up the information that we've parsed a naked lifetime label token through the various layers of parse_expr, since we just return an expression (and doing so would be a very heavy change). We can't bubble down the expectation that we _might_ be currently parsing something that looks like a bad turbofish either, because we already parsed the `<` of the turbofish as a regular operator expression.

This example, which ICEs, might help demonstrate something that locally looks like a bad turbofish, but which parses (incorrectly) today, since we only delay a bug when we parse that bad lifetime `'a`. I am just adding emitting an error back, since we'll have emitted at least one (which might be redudant, in the case of a bad turbofish) to prevent any ICEs.
```rust
fn f(_a1: bool, _a2: bool) {
    let a = 1i32;
    let b = 1i32;
    f(a < 'a, b > (1))
}
```

**Pull request merged**